jobbergate-api/jobbergate_api/s3_manager.py

```diff
@@ -1,37 +1,157 @@
+        Delete a file from the client associated to the given key.
+        """
+        try:
+            del self.client[key]
+        except KeyError:
+            raise KeyError(f"No such key: {key}")
+
+    def get(self, key: str) -> typing.IO:
+        """
+        Get a file from the client associated to the given key.
+        """
+        try:
+            return self.client[key]
+        except KeyError:
+            raise KeyError(f"No such key: {key}")
+
+
+class S3Client(BucketClientBase):
+    """
+    S3 client.
+    """
+
+    def __init__(self):
+        """
+        Initialize an s3 client.
         """
         self.bucket_name = settings.S3_BUCKET_NAME
-        self.key_template = "applications/{app_id}/jobbergate.tar.gz"
         self.s3_client = boto3.client(
             "s3",
             endpoint_url=settings.S3_ENDPOINT_URL,
             aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
             aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
         )
 
-    def _get_key(self, app_id):
+    def put(self, file: typing.IO, key: str) -> None:
+        """
+        Upload a file to the client for the given key.
+        """
+        self.s3_client.put_object(
+            Body=file,
+            Bucket=self.bucket_name,
+            Key=key,
+        )
+
+    def delete(self, key: str) -> None:
+        """
+        Delete a file from the client associated to the given key.
+        """
+        try:
+            self.s3_client.delete_object(Bucket=self.bucket_name, Key=key)
+        except self.s3_client.exceptions.NoSuchBucket:
+            raise KeyError(f"No such bucket: {self.bucket_name}")
+        except self.s3_client.exceptions.NoSuchKey:
+            raise KeyError(f"No such key: {key}")
+
+    def get(self, key: str) -> typing.IO:
+        """
+        Get a file from the client associated to the given key.
+        """
+        # TODO: handle client errors and raise KeyError when applicable
+        return self.s3_client.get_object(Bucket=self.bucket_name, Key=key)
+
+
+class S3Manager:
+    """
+    Provide a class for managing connection to an S3 bucket.
+
+    Prepares the bucket_name and s3 key template from project settings. Also initialize an s3
+    client. Provide methods as thin wrappers around s3 calls but provide the mappings for buckets and keys.
+    """
+
+    def __init__(
+        self,
+        directory_name: str = "applications",
+        filename="jobbergate.tar.gz",
+        client: BucketClientBase = None,
+    ):
+        """
+        Initialize the S3Manager class instance.
+        """
+        self.key_template = directory_name + "/{app_id}/" + filename
+
+        if client is None:
+            client = S3Client()
+
+        self.client = client
+
+    def _get_key(self, app_id: str) -> str:
         """
-        Get an s3 key based upon the app_id. If app_id is falsey, throw an exception.
+        Get an s3 key based upon the app_id. If app_id is empty, throw an exception.
```

Comment:
Well, in this case....perhaps it's better as `empty`. Since the type-hint suggests that `app_id` should only be a string and because only an empty string resolves as `False`, this is more narrowly correct than saying `falsey`.

Reply:
Ok, no need to change.